### ISSUE TITLE
[K8s] Storage mounting requires privileged containers

### ISSUE THREAD

**Issue by romilbhardwaj** [skypilot-org/skypilot#2356]
Our storage mounting (goofys, gcsfuse) is based on FUSE mounts. Using FUSE mounts in containers requires running containers with `SYS_ADMIN` [capabilities](https://stackoverflow.com/questions/48402218/fuse-inside-docker), which may not be permitted on many clusters and may be seen as a security risk.

We currently create privileged containers:

https://github.com/skypilot-org/skypilot/blob/2b2a158c6730f605ca61bc8c4b051b278e3b7474/sky/templates/kubernetes-ray.yml.j2#L172-L175

We may need to reconsider this and allow only `COPY` or `C_SYNC` modes for Storages attached to SkyPilot clusters on Kubernetes. Creating this ticket to keep track of this issue.

**Comment by romilbhardwaj** [comment]
After some initial exploration, simply using `SYS_ADMIN` does not work:
```
          securityContext:
            capabilities:
              add:
              - SYS_ADMIN
```

```
(base) sky@test-2ea4-ray-head:~$ goofys -f romil-sky-nfs2 wow
2023/11/06 21:05:30.852334 main.INFO bucket romil-sky-nfs2
2023/11/06 21:05:30.852568 main.INFO prefix
2023/11/06 21:05:30.852591 main.INFO testing object um81ismkx2m26aiy4zd858u5sw8d1okp
2023/11/06 21:05:31.104623 s3.INFO Switching from region 'us-east-1' to 'us-west-2'
2023/11/06 21:05:31.324170 main.FATAL Mounting file system: Mount: mount: running fusermount: exit status 1

stderr:
fusermount: failed to open /dev/fuse: Operation not permitted
```

We may need to do something like [this](https://github.com/kubernetes/kubernetes/issues/7890#issuecomment-766088805).

**Comment by romilbhardwaj** [comment]
After some investigation:

* [fuse-csi driver](https://github.com/pfnet-research/meta-fuse-csi-plugin) works for mounting a single endpoint by simply adding a volumeMount for the fuse device exposed on all nodes through a daemonset. However, only one process can consume the /dev/fuse device at a time, limiting the ability to run multiple mounts together.
* [https://github.com/smarter-project/smarter-device-manager](smarter-device-manager) works by exposing /dev/fuse as an extended resource. We can expose it with a large capacity (say 9999) and then request the device in our resource request. This works even with multiple devices, and is the choice we will be going ahead with. 

An example YAML to deploy device-manager
```
apiVersion: v1
kind: Namespace
metadata:
  name: device-manager
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: smarter-device-manager
  namespace: device-manager
data:
  conf.yaml: |
    - devicematch: ^fuse$
      nummaxdevices: 20
---
# https://gitlab.com/arm-research/smarter/smarter-device-manager/-/blob/master/smarter-device-manager-ds.yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: smarter-device-manager
  namespace: device-manager
  labels:
    name: smarter-device-manager
    role: agent
spec:
  selector:
    matchLabels:
      name: smarter-device-manager
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        name: smarter-device-manager
      annotations:
        node.kubernetes.io/bootstrap-checkpoint: "true"
    spec:
      ## kubectl label node pike5 smarter-device-manager=enabled
      # nodeSelector:
      #   smarter-device-manager : enabled
      # priorityClassName: "system-node-critical"
      hostname: smarter-device-management
      hostNetwork: true
      dnsPolicy: ClusterFirstWithHostNet
      containers:
      - name: smarter-device-manager
        image: registry.gitlab.com/arm-research/smarter/smarter-device-manager:v1.1.2
        imagePullPolicy: IfNotPresent
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop: ["ALL"]
        resources:
          limits:
            cpu: 100m
            memory: 15Mi
          requests:
            cpu: 10m
            memory: 15Mi
        volumeMounts:
          - name: device-plugin
            mountPath: /var/lib/kubelet/device-plugins
          - name: dev-dir
            mountPath: /dev
          - name: sys-dir
            mountPath: /sys
          - name: config
            mountPath: /root/config
      volumes:
        - name: device-plugin
          hostPath:
            path: /var/lib/kubelet/device-plugins
        - name: dev-dir
          hostPath:
            path: /dev
        - name: sys-dir
          hostPath:
            path: /sys
        - name: config
          configMap:
            name: smarter-device-manager

# Use in pod with:
#          resources:
#            limits:
#              smarter-devices/fuse: 1
#              memory: 512Mi
#            requests:
#              smarter-devices/fuse: 1
#              cpu: 10m
#              memory: 50Mi
```

**Comment by romilbhardwaj** [comment]
Fixed with #3443.
